agent/consul/server.go

```diff
@@ -1250,7 +1250,7 @@ func (s *Server) setupExternalGRPC(config *Config, logger hclog.Logger, backend
 	registry := resource.NewRegistry()
 
 	if s.config.DevMode {
-		demo.Register(registry)
+		demo.Register(registry, nil)
```

Comment:
I didn't want to change the signature to `Register` to pass in ACLs, but the `Registration` object is effectively immutable post-registration since it passed around by value. Open to better suggestions for this as our tests scale.

Reply:
That's a really good point! Ideally, I think the `demo` package should have a realistic set of ACL hooks (e.g. `demo.v1.artist:write`, `demo.v1.album:read`), and rather than mocking out the hooks in our tests, we should construct authorizers using real policy HCL ([example](https://github.com/hashicorp/consul/blob/0582f137c53f6200575eea5a2aef8257e0f0e805/agent/grpc-external/testutils/acl.go#L46-L63)).

Unfortunately, our current ACL system (like most of our stack) requires _manually_ adding new resource types and stanzas - something we'll need to address in the future, but out of scope for our current work. 😞 

I wonder if, for now, we could fake it using K/V permissions? For example, `demo.v1.artist:write` could be implemented as:

```Go
func(authz acl.Authorizer, res *pbresource.Resource) error {
	key := fmt.Sprintf("resource/%s/%s", resource.ToGVK(res.ID.Type), res.ID.Name)
	return authz.ToAllowAuthorizer().KeyWriteAllowed(key, &acl.AuthorizerContext{})
}
```

Then, for tests, the `demo` package could expose a policy like this:

```Go
const ArtistV1WritePolicy = `key_prefix "resource/demo.v1.artist/" { policy = "write" }`
```

Which we could use to construct a policy authorizer in our endpoint tests, and we'd no longer need that second argument to `demo.Register`.

Slightly whacky idea. What do you think? 😄

Reply:
Sounds worth trying. I did this for the `Read` tests only in https://github.com/hashicorp/consul/pull/16842/commits/45352a443123a65c783a9d4d17298e79cfaadec1. Let me know what you think.

Reply:
It's a little odd, for sure, but I think it actually works quite well as a stand-in until our ACL system can be more modularly extended (or support generic resource permissions). What do you think?

If we keep it, we should probably add some comments explaining why we're using K/V rules in this way. I also notice that the `Read` tests exercise the cases where `Registration.ACLs` is nil or missing hooks, IMO we should cover this in the registry's unit tests instead (so we can go ahead and remove that second argument to `demo.Register`).

Reply:
Yeah, I think it is worth keeping as a workaround until we're further down the road and it does the job for now. I'll go ahead and apply this to the other endpoints.